### PR TITLE
feat: ✨ ollama streaming mode

### PR DESCRIPTION
**Pull request by philippart-s**
PR to try to add the streaming feature to Ollama provider through the extension 😉 

First commits are the setup of the the feature 🏗️.

- Closes: #501

### CODE REVIEW

File: ollama/deployment/src/main/java/io/quarkiverse/langchain4j/ollama/deployment/OllamaProcessor.java [hunk 1]
```diff
@@ -1,7 +1,6 @@
 package io.quarkiverse.langchain4j.ollama.deployment;
 
-import static io.quarkiverse.langchain4j.deployment.LangChain4jDotNames.CHAT_MODEL;
-import static io.quarkiverse.langchain4j.deployment.LangChain4jDotNames.EMBEDDING_MODEL;
+import static io.quarkiverse.langchain4j.deployment.LangChain4jDotNames.*;
```
Comment (geoand) [inline ollama/deployment/src/main/java/io/quarkiverse/langchain4j/ollama/deployment/OllamaProcessor.java:3]: We generally avoid star imports

File: ollama/deployment/src/main/java/io/quarkiverse/langchain4j/ollama/deployment/OllamaProcessor.java [hunk 2]
```diff
@@ -66,6 +65,30 @@ void generateBeans(OllamaRecorder recorder,
...
+        for (var selected : selectedEmbedding) {
+            if (PROVIDER.equals(selected.getProvider())) {
+                String modelName = selected.getModelName();
+                var builder = SyntheticBeanBuildItem
+                        .configure(EMBEDDING_MODEL)
+                        .setRuntimeInit()
+                        .defaultBean()
+                        .unremovable()
+                        .scope(ApplicationScoped.class)
+                        .supplier(recorder.embeddingModel(config, modelName));
```
Comment (geoand) [inline ollama/deployment/src/main/java/io/quarkiverse/langchain4j/ollama/deployment/OllamaProcessor.java:89]: Was an embedding model added?

File: ollama/runtime/pom.xml [hunk 3]
```diff
@@ -26,7 +26,36 @@
             <artifactId>quarkus-langchain4j-core</artifactId>
             <version>${project.version}</version>
         </dependency>
-
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
```
Comment (geoand) [inline ollama/runtime/pom.xml:30]: Why this change?

**Review by geoand — COMMENTED** [review]
Thanks for the start!

I added some comments

File: ollama/deployment/src/main/java/io/quarkiverse/langchain4j/ollama/deployment/OllamaProcessor.java (hunk 1 above)
Comment (philippart-s) [reply ollama/deployment/src/main/java/io/quarkiverse/langchain4j/ollama/deployment/OllamaProcessor.java:3]: yes ou are right, I think it's my IDE.
I had to switch to Intellij because my VSCode was crashing when I try to use it to dev on the extension. So my settings are not quite perfect on my Intellij 😉

File: ollama/deployment/src/main/java/io/quarkiverse/langchain4j/ollama/deployment/OllamaProcessor.java (hunk 2 above)
Comment (philippart-s) [reply ollama/deployment/src/main/java/io/quarkiverse/langchain4j/ollama/deployment/OllamaProcessor.java]: wrong copy/past, thanks I deleted it.

File: ollama/runtime/pom.xml (hunk 3 above)
Comment (philippart-s) [reply ollama/runtime/pom.xml]: I needed it for one of my tests, but it's no longer in use.
I've deleted it (and I'll see if I need some classes from this dependency later 😉).

Comment (philippart-s) [reply ollama/runtime/pom.xml]: ➡️ https://github.com/quarkiverse/quarkus-langchain4j/pull/522/commits/35a077bd3c62352daa97f70c493a2e2d15f0c265

File: ollama/deployment/src/main/java/io/quarkiverse/langchain4j/ollama/deployment/OllamaProcessor.java (hunk 1 above)
Comment (philippart-s) [reply ollama/deployment/src/main/java/io/quarkiverse/langchain4j/ollama/deployment/OllamaProcessor.java]: ➡️ https://github.com/quarkiverse/quarkus-langchain4j/pull/522/commits/efc0de2213cdc181d26d9377a46890c403b0642b

File: ollama/runtime/pom.xml (hunk 3 above)
Comment (philippart-s) [reply ollama/runtime/pom.xml]: @geoand after remove it i remember (or Maven remember to me 😉) why I've needed the add the dependency: it was to import the class `dev.langchain4j.model.ollama.OllamaStreamingChatModel`.


If I well understood, for the supplier I need a builder and this builder is in this class.

Comment (philippart-s) [reply ollama/runtime/pom.xml]: I confess that at this stage, I don't understand yet why, for some models, some classes are used from langchain4j and other rewritten with duplicated code.

Other thinks I need to understand, for example, is why for some models the langchain4j client is used or a new one is created (like Mistral for example).

I'm trying to understand the architecture of the code, and I think I'm going to make some mistakes, sorry.

Comment (geoand) [reply ollama/runtime/pom.xml]: The general idea was that we try to reuse upstream bit, but for some integrations the upstream code is so mimimal that it just does not bring any benefit to do so.
Such is the case of the Ollama extension

**Comment by geoand** [comment]
Thanks, this is a nice start! The most important part now is the implementation of `OllamaStreamingChatLanguageModel`.

File: ollama/runtime/pom.xml (hunk 3 above)
Comment (philippart-s) [reply ollama/runtime/pom.xml]: yes, I begin to understand this and I agree with you 😉

**Comment by philippart-s** [comment]
> Thanks, this is a nice start! The most important part now is the implementation of `OllamaStreamingChatLanguageModel`.

Yes, I've started the implementation, I need to correct one or two mistakes but it's coming along!

**Comment by andreadimaio** [comment]
@philippart-s I think that this [link](https://github.com/ollama/ollama/blob/main/docs/api.md#generate-a-chat-completion) can help you to understand how to create the REST API call for the streaming. About the code implementation,  you might want to take a look at what I did for the BAM module [here ](https://github.com/quarkiverse/quarkus-langchain4j/blob/5f8e2e59ce8b7c1fd32e4f733690bd587a45fdce/bam/runtime/src/main/java/io/quarkiverse/langchain4j/bam/BamRestApi.java#L52-L56)and [here](https://github.com/quarkiverse/quarkus-langchain4j/blob/5f8e2e59ce8b7c1fd32e4f733690bd587a45fdce/bam/runtime/src/main/java/io/quarkiverse/langchain4j/bam/BamStreamingChatModel.java#L54-L113) (I hope this can help you).

**Comment by philippart-s** [comment]
@geoand : I've a first version that I want to test but when I run the integration tests (for example https://github.com/quarkiverse/quarkus-langchain4j/tree/main/integration-tests/ollama, I've this error: `2024-04-30 15:13:06,715 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.IllegalStateException: No config found for interface io.quarkiverse.langchain4j.runtime.aiservice.ChatMemoryConfig`

I run the `quarkus dev` command.
Is there a tips to run the examples in the integration-test folder?

**Comment by philippart-s** [comment]
> 🎉
> 
> I would first build the entire project and then do `mvn quarkus:dev` in the `integration-tests/ollama` directory.

it's what I did:

```bash
➜ cd quarkus-langchain4j/integration-tests/ollama 

➜ mvn quarkus:dev
[INFO] Scanning for projects...
[INFO] 
[INFO] --< io.quarkiverse.langchain4j:quarkus-langchain4j-integration-test-ollama >--
[INFO] Building Quarkus LangChain4j - Integration Tests - Ollama 999-SNAPSHOT
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- quarkus:3.8.2:dev (default-cli) @ quarkus-langchain4j-integration-test-ollama ---
[INFO] Invoking enforcer:3.4.1:enforce (enforce-java-version) @ quarkus-langchain4j-integration-test-ollama
[INFO] Rule 0: org.apache.maven.enforcer.rules.BannedRepositories passed
[INFO] Rule 1: org.apache.maven.enforcer.rules.version.RequireJavaVersion passed
[INFO] Invoking enforcer:3.4.1:enforce (enforce-maven-version) @ quarkus-langchain4j-integration-test-ollama
[INFO] Rule 0: org.apache.maven.enforcer.rules.version.RequireMavenVersion passed
[INFO] Invoking sundr:0.103.1:generate-bom (default) @ quarkus-langchain4j-integration-test-ollama
[INFO] Invoking buildnumber:3.2.0:create (get-scm-revision) @ quarkus-langchain4j-integration-test-ollama
[INFO] Executing: /bin/sh -c cd '/Users/stef/Dev/quarkus-langchain4j/integration-tests/ollama' && 'git' 'rev-parse' '--verify' 'HEAD'
[INFO] Working directory: /Users/stef/Dev/quarkus-langchain4j/integration-tests/ollama
[INFO] Storing buildNumber: efc0de2213cdc181d26d9377a46890c403b0642b at timestamp: 1714568697753
[INFO] Executing: /bin/sh -c cd '/Users/stef/Dev/quarkus-langchain4j/integration-tests/ollama' && 'git' 'symbolic-ref' 'HEAD'
[INFO] Working directory: /Users/stef/Dev/quarkus-langchain4j/integration-tests/ollama
[INFO] Storing scmBranch: ollama-streaming
[INFO] Invoking formatter:2.23.0:format (format-sources) @ quarkus-langchain4j-integration-test-ollama
[INFO] Processed 1 files in 259ms (Formatted: 0, Skipped: 1, Unchanged: 0, Failed: 0, Readonly: 0)
[INFO] Invoking impsort:1.9.0:sort (sort-imports) @ quarkus-langchain4j-integration-test-ollama
[INFO] Processed 1 files in 00:00.005 (Already Sorted: 1, Needed Sorting: 0)
[INFO] Invoking resources:3.3.1:resources (default-resources) @ quarkus-langchain4j-integration-test-ollama
[INFO] Copying 1 resource from src/main/resources to target/classes
[INFO] Invoking compiler:3.12.1:compile (default-compile) @ quarkus-langchain4j-integration-test-ollama
[INFO] Nothing to compile - all classes are up to date.
[INFO] Invoking resources:3.3.1:testResources (default-testResources) @ quarkus-langchain4j-integration-test-ollama
[INFO] skip non existing resourceDirectory /Users/stef/Dev/quarkus-langchain4j/integration-tests/ollama/src/test/resources
[INFO] Invoking compiler:3.12.1:testCompile (default-testCompile) @ quarkus-langchain4j-integration-test-ollama
[INFO] No sources to compile
[WARNING] [io.quarkus.bootstrap.devmode.DependenciesFilter] Live reload was disabled for the following project artifacts:
- io.quarkiverse.langchain4j:quarkus-langchain4j-ollama:999-SNAPSHOT
- io.quarkiverse.langchain4j:quarkus-langchain4j-core:999-SNAPSHOT
- io.quarkiverse.langchain4j:quarkus-langchain4j-core-runtime-spi:999-SNAPSHOT
- io.quarkiverse.langchain4j:quarkus-langchain4j-core-deployment:999-SNAPSHOT
- io.quarkiverse.langchain4j:quarkus-langchain4j-ollama-deployment:999-SNAPSHOT
The artifacts above appear to be either dependencies of non-reloadable application dependencies or Quarkus extensions
Listening for transport dt_socket at address: 5005
2024-05-01 15:05:00,120 INFO  [io.qua.dep.dev.IsolatedDevModeMain] (main) Attempting to start live reload endpoint to recover from previous Quarkus startup failure
2024-05-01 15:05:00,335 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.IllegalStateException: No config found for interface io.quarkiverse.langchain4j.runtime.aiservice.ChatMemoryConfig
        at io.quarkus.deployment.ExtensionLoader.loadStepsFrom(ExtensionLoader.java:186)
        at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:107)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:330)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:251)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:60)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:112)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:433)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:55)
        at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:138)
        at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:93)
        at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:131)
        at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:62)
```

I have this error for all tests in the `integration-tests`, I should do something wrong but I don't understand what.

**Comment by geoand** [comment]
Weird... I've never seen that happen and it works fine for me (and CI)

**Comment by philippart-s** [comment]
I'll check again all my configuration to see if I have an issue on my local configuration.

**Comment by geoand** [comment]
Would you like to push what you have so I can check it out locally?

**Comment by philippart-s** [comment]
after a fresh build it's working 🎉
I'm going to push my current code but I think it's not perfect (I haven't dev the number of tokens yet).

**Comment by geoand** [comment]
Thanks a lot @philippart-s! 

I will take your code and finish the implementation soon.

**Comment by geoand** [comment]
No need to apologize!

Thanks a lot for getting the ball rolling on this one!

I will ping you when I've completed the PR :)

**Comment by geoand** [comment]
The problem with the exception you are seeing turns out to be weirder than I thought...

All we can do for the time being is workaround it as I have done.

Do you mind checking if things work for you with the latest version (I've force pushed to your branch so you'll have to be careful when pulling)

**Comment by philippart-s** [comment]
yes, I'm going to try it as soon as I have a bit of time in my day 😉

**Comment by philippart-s** [comment]
@geoand I tested with my app and it seems ok 👌.
Thanks to completed the code to made it correct!

Just for my information, for the next PR, what type of configuration for the code formatter is used? (to avoid the maven verify error on the CI 😅)

**Comment by geoand** [comment]
Thanks for checking!

All I do is `mvn install -f ollama` and the build does the formatting automatically

**Comment by adriens** [comment]
:clap:
